**Import Required Libraries**

In [65]:
import pandas as pd
import numpy as np

**pandas** is used for data manipulation and analysis

**numpy** is used for numerical operations and handling arrays

**Load Dataset**

In [66]:
data = pd.read_csv('daily_weather.csv')

**Dataset Overview**

The dataset contains daily weather observations consisting of 4081 records and multiple weather-related variables. Each record represents the weather conditions for a specific day and includes information such as temperature, rainfall, wind speed, and derived temporal features.

The dataset includes temperature measurements (maximum, minimum, and mean), rainfall indicators, wind speed, and time-based attributes such as year, month, weekday, and season. Additional engineered features like temperature range and average temperature further enhance the dataset for deeper analysis.

Overall, the dataset provides a comprehensive view of weather patterns over several years, making it suitable for climate analysis, trend identification, and predictive modeling.

| Column         | Description                                        |
| -------------- | -------------------------------------------------- |
| date           | Date of the weather observation                    |
| temp_max       | Maximum temperature recorded for the day           |
| temp_min       | Minimum temperature recorded for the day           |
| temp_mean      | Average daily temperature                          |
| precipitation  | Amount of rainfall recorded                        |
| rain           | Rainfall measurement indicator                     |
| wind_speed_max | Maximum wind speed recorded                        |
| rainflag       | Boolean flag indicating presence of rain           |
| temp_range     | Difference between maximum and minimum temperature |
| temp_avg       | Average temperature value                          |
| month          | Month extracted from the date                      |
| season         | Season category derived from the date              |


**Scope of the Dataset**

The dataset supports several types of weather-related analysis, including:

**Temperature Analysis**

Maximum, minimum, and average temperature trends

Identification of hot days and temperature extremes

**Rainfall Analysis**

Frequency of rainfall events

Rain vs clear sky comparison

**Wind Analysis**

Identification of windy days

Wind speed distribution

**Seasonal Analysis**

Weather patterns across four seasons

Temperature and rainfall variations by season

**Extreme Weather Detection**

Identification of days with high temperature and strong winds

The dataset can also be used for:

Exploratory Data Analysis (EDA)

Climate trend analysis

Predictive modeling and machine learning

Weather forecasting models

Because the dataset contains temporal, numerical, and categorical features, it supports statistical analysis, visualization, and predictive modeling.

**Limitations of the Dataset**

Despite its usefulness, the dataset has certain limitations:

1. The dataset includes only a limited number of weather parameters, mainly temperature, rainfall, and wind speed.

2. Important meteorological factors such as pressure, humidity, cloud cover, and solar radiation are not included.

3. The dataset represents a specific geographical location, which limits the ability to generalize findings to other regions.

4. Rainfall information is relatively simplified, with a rain flag indicating rain or clear sky conditions.

These limitations may restrict the depth of climate modeling and large-scale meteorological analysis.

**Dataset Summary Statistics**

In [67]:
data.describe()

,temp_max,temp_min,temp_mean,precipitation,rain,wind_speed_max
count,4081.000000,4081.000000,4081.000000,4081.000000,4081.000000,4081.000000
mean,31.085665,21.296373,25.841338,2.384220,2.384220,15.443396
std,3.644176,3.267929,3.128517,6.006201,6.006201,5.494186
min,20.100000,10.900000,17.300000,0.000000,0.000000,5.100000
25%,28.500000,18.900000,23.700000,0.000000,0.000000,11.200000
50%,30.000000,22.100000,25.400000,0.000000,0.000000,14.100000
75%,33.400000,23.400000,27.600000,1.600000,1.600000,18.400000
max,41.500000,29.800000,35.600000,82.700000,82.700000,37.900000



**Temperature**

The average maximum temperature is approximately 31.09 °C, indicating generally warm daytime conditions.

The average minimum temperature is about 21.30 °C, suggesting moderate night temperatures.

The average mean temperature across the dataset is 25.84 °C, reflecting a predominantly warm climate overall.

Maximum temperatures range from 20.1 °C to 41.5 °C, showing that the region occasionally experiences very hot days.

Minimum temperatures vary from 10.9 °C to 29.8 °C, indicating some cooler periods during certain seasons.

The interquartile range shows that 50% of maximum temperatures lie between 28.5 °C and 33.4 °C, suggesting most days experience moderately high temperatures.

**Rainfall / Precipitation**

The average precipitation is 2.38 mm, indicating generally low daily rainfall across the dataset.

The median precipitation value is 0 mm, meaning that more than half of the days experienced no rainfall.

The maximum precipitation recorded is 82.7 mm, indicating occasional heavy rainfall events.

The 75th percentile is 1.6 mm, suggesting that even during wetter days, rainfall levels are typically relatively light.

**Wind Speed**

The average maximum wind speed is 15.44 km/h, indicating moderate wind conditions overall.

Wind speeds range from 5.1 km/h to 37.9 km/h, showing considerable variability in wind intensity.

The median wind speed is 14.1 km/h, suggesting that typical daily wind conditions are moderate rather than extreme.

The upper quartile value of 18.4 km/h indicates that about 25% of days experience relatively stronger winds.

**Converting Columns into Series**

In [68]:
temp_mean = data["temp_mean"]
precipitation = data["precipitation"]
rain = data["rain"]

Extracts individual columns from the dataset as pandas Series objects.

**Inspecting the Data**

In [69]:
print("First 5 values of mean temperature:\n", temp_mean.head())

print("\nLast 5 values of mean temperature:\n", temp_mean.tail())

print("\nRandom sample of mean temperature:\n", temp_mean.sample())

print("\nTotal number of mean temperature records:", temp_mean.size)

print("\nData type of mean temperature:", temp_mean.dtype)

print("\nIndex of mean temperature column:", temp_mean.index)

print("\nValues of mean temperature column:\n", temp_mean.values)

First 5 values of mean temperature:
 0    23.2
1    23.1
2    22.3
3    21.7
4    22.1
Name: temp_mean, dtype: float64

Last 5 values of mean temperature:
 4076    26.4
4077    27.3
4078    28.3
4079    28.8
4080    28.2
Name: temp_mean, dtype: float64

Random sample of mean temperature:
 3652    24.5
Name: temp_mean, dtype: float64

Total number of mean temperature records: 4081

Data type of mean temperature: float64

Index of mean temperature column: RangeIndex(start=0, stop=4081, step=1)

Values of mean temperature column:
 [23.2 23.1 22.3 ... 28.3 28.8 28.2]


**Counting Rainy and Non-Rainy Days**

In [70]:
data["rainflag"] = data["precipitation"] > 0
data["rainflag"].value_counts()
data["rainflag"].map({True:'rain',False:'clear sky'}).value_counts()

rainflag
clear sky    2237
rain         1844
Name: count, dtype: int64

The dataset contains 4081 total weather records, divided into two rain conditions: clear sky and rain.

Clear sky days account for **2237** observations, representing the majority of the dataset.

Rainy days account for **1844** observations, indicating that rainfall occurs fairly frequently but slightly less often than dry conditions.

The distribution shows that approximately **55%** of the days experienced clear skies, while about **45%** of the days had rainfall.

This relatively balanced distribution suggests the dataset includes both dry and rainy conditions, making it useful for weather pattern analysis, rainfall prediction, and climate trend studies.

The presence of a large number of both classes also ensures the dataset is suitable for machine learning models, particularly for rainfall classification or weather forecasting tasks.

**Boolean Indexing**

In [71]:
hotdays = data[data["temp_max"]>30]
len(hotdays)

2038

In [72]:
windydays = data[data["wind_speed_max"]>10]
len(windydays)

3551

**Hot Days**

A total of 2038 days recorded maximum temperatures above 30 °C, indicating a significant number of hot weather conditions in the dataset.

This suggests that nearly half of the recorded days experience high daytime temperatures, reflecting a generally warm climate pattern.

Frequent hot days may indicate seasonal heat periods or prolonged warm weather conditions.

**Windy Days**

There are 3551 days where the maximum wind speed exceeds 10 km/h, indicating that windy conditions occur quite frequently.

This means that most days experience noticeable wind activity, suggesting consistent atmospheric movement in the region.

The high number of windy days may influence temperature distribution, cloud movement, and rainfall patterns.

**Overall Insight**

The dataset shows a combination of frequent warm temperatures and regular wind activity, which can significantly influence local weather dynamics and seasonal climate behavior.

**Inspecting Dataset Structure**

In [73]:
data.shape

(4081, 8)

The dataset contains 4081 records (rows) and 8 variables (columns), indicating a reasonably large dataset suitable for weather analysis.

The columns present in the dataset are:

date – represents the temporal information for each weather observation.

temp_max – maximum temperature recorded for the day.

temp_min – minimum temperature recorded for the day.

temp_mean – average temperature of the day.

precipitation – amount of rainfall recorded.

rain – rainfall indicator/measurement.

wind_speed_max – maximum wind speed recorded for the day.

rainflag – categorical label indicating clear sky or rain condition.

In [74]:
data.columns

Index(['date', 'temp_max', 'temp_min', 'temp_mean', 'precipitation', 'rain',
       'wind_speed_max', 'rainflag'],
      dtype='object')

**Numerical Columns:** 6

These include variables such as temperature, precipitation, and wind speed, which contain continuous numeric values used for statistical analysis and modeling.

**Boolean Column:** 1

The dataset contains one boolean variable (rain), which represents whether rainfall occurred or not.

**Text/Date Column:** 1

The date column stores temporal information, representing the specific day of each weather observation.

**Creating New Features**

Temperature Range

In [75]:
data["temp_range"] = data['temp_max'] - data['temp_min']

Average Temperature

In [76]:
data["temp_avg"] = (data['temp_max'] + data['temp_min']) / 2

**Column Selection**


In [77]:
data['date']

0       2015-01-01
1       2015-01-02
2       2015-01-03
3       2015-01-04
4       2015-01-05
           ...    
4076    2026-02-28
4077    2026-03-01
4078    2026-03-02
4079    2026-03-03
4080    2026-03-04
Name: date, Length: 4081, dtype: object

In [78]:
data[["date", "rainflag"]]

,date,rainflag
0,2015-01-01,True
1,2015-01-02,False
2,2015-01-03,True
3,2015-01-04,True
4,2015-01-05,False
...,...,...
4076,2026-02-28,False
4077,2026-03-01,False
4078,2026-03-02,False
4079,2026-03-03,False


**Filtering Data**

In [79]:
extreme_weather = data[(data["temp_max"]>30) & (data['wind_speed_max']>10)]
len(extreme_weather)

1829

**Working with Date and Time**

In [80]:
import datetime

In [81]:
data['date'] = pd.to_datetime(data['date'])
data["month"] = data['date'].dt.month

**Creating Seasonal Categories**

In [82]:
def get_seasons_month(month):
    if month in [12,1,2]:
        return "winter"
    elif month in [3,4,5]:
        return "summer"
    elif month in [6,7,8]:
        return "monsoon"
    else:
        return "autumn"
data["season"] = data["month"].apply(get_seasons_month)

In [83]:
data.head()

,date,temp_max,temp_min,temp_mean,precipitation,rain,wind_speed_max,rainflag,temp_range,temp_avg,month,season
0,2015-01-01,28.2,18.8,23.2,0.2,0.2,15.7,True,9.4,23.50,1,winter
1,2015-01-02,27.5,19.5,23.1,0.0,0.0,12.8,False,8.0,23.50,1,winter
2,2015-01-03,26.5,19.6,22.3,4.0,4.0,10.3,True,6.9,23.05,1,winter
3,2015-01-04,25.6,19.5,21.7,0.6,0.6,9.3,True,6.1,22.55,1,winter
4,2015-01-05,27.6,17.6,22.1,0.0,0.0,13.6,False,10.0,22.60,1,winter


**Converting Data Types**

In [84]:
data["season"] = data["season"].astype("category")
data["rainflag"] = data["rainflag"].astype("category")

**Data Cleaning**

Finding missing values

In [85]:
data.isnull().sum()

date              0
temp_max          0
temp_min          0
temp_mean         0
precipitation     0
rain              0
wind_speed_max    0
rainflag          0
temp_range        0
temp_avg          0
month             0
season            0
dtype: int64

**Checking Duplicate Records**

In [86]:
data.duplicated().sum()

np.int64(0)

**Sorting Data by Maximum Temperature**

In [87]:
data.sort_values(["temp_max"], ascending=False)

,date,temp_max,temp_min,temp_mean,precipitation,rain,wind_speed_max,rainflag,temp_range,temp_avg,month,season
140,2015-05-21,41.5,28.5,34.6,0.0,0.0,18.3,False,13.0,35.00,5,summer
477,2016-04-22,41.4,29.8,35.6,0.0,0.0,11.7,False,11.6,35.60,4,summer
3411,2024-05-04,41.4,24.9,33.1,0.0,0.0,14.7,False,16.5,33.15,5,summer
480,2016-04-25,41.2,26.2,34.3,0.0,0.0,16.3,False,15.0,33.70,4,summer
3413,2024-05-06,40.8,25.9,33.6,0.0,0.0,15.3,False,14.9,33.35,5,summer
...,...,...,...,...,...,...,...,...,...,...,...,...
631,2016-09-23,22.9,21.8,22.5,46.1,46.1,19.1,True,1.1,22.35,9,autumn
1487,2019-01-27,22.7,17.9,19.2,21.3,21.3,15.3,True,4.8,20.30,1,winter
1447,2018-12-18,22.7,15.6,18.4,1.9,1.9,18.9,True,7.1,19.15,12,winter
1865,2020-02-09,21.5,18.5,19.7,17.0,17.0,15.5,True,3.0,20.00,2,winter


**Duplicate and Missing Data Check**

Duplicate Records: 0

The dataset contains no duplicate rows, indicating that each weather observation is unique.

Missing Values: 0

The dataset also contains no null or missing values across any columns.

**Insight**

The absence of duplicate and missing data indicates strong data quality. This reduces the need for additional preprocessing steps such as data cleaning, duplicate removal, or missing value imputation, making the dataset ready for analysis, visualization, and machine learning applications.

**Extracting More Date Features**

In [88]:
data["year"] = data["date"].dt.year

In [89]:
data["weekday"] = data["date"].dt.weekday

In [90]:
data["day"] = data["date"].dt.day

In [91]:
data["dayname"] = data["date"].dt.day_name()

**Checking Number of Unique Seasons**

In [92]:
data["season"].nunique()

4

**Renaming Columns**

In [93]:
data.rename(columns={"precipitation":"humidity"},inplace=True)

**Feature Engineering and Data Transformation**

Several new features were extracted from the original dataset to enhance analysis and improve the dataset’s usefulness for modeling and visualization.

**New Features Created**

Year – Extracted from the date column to analyze long-term climate trends.

Month – Helps identify monthly weather patterns.

Season – Categorizes weather into seasonal groups such as winter, summer, monsoon, etc. for seasonal analysis.

Weekday – Numerical representation of the day of the week.

Day – Represents the day of the month.

Dayname – Provides the actual weekday name (e.g., Monday, Tuesday).

**Derived Temperature Features**

temp_range – Calculated as the difference between maximum and minimum temperature, representing daily temperature variation.

temp_avg – Average temperature value used for smoother temperature analysis.

**Column Transformation**

The column precipitation was renamed to humidity to improve interpretability and readability within the dataset.

**Saving Cleaned Dataset**

In [94]:
data.to_csv("Clean_weather.csv")

**Dataset Integrity and Quality Evaluation**

The dataset demonstrates high data integrity and quality, making it reliable for analysis.

1. **Data Completeness**

The dataset contains no missing values across all columns.

Each record includes complete weather information for the given date.

2. **Duplicate Records**

No duplicate entries were found in the dataset, ensuring each observation represents a unique day.

3. **Data Consistency**

Numerical variables such as temperature, precipitation, and wind speed are stored in consistent numeric formats.

Boolean variables correctly represent rainfall occurrence.

Temporal information is properly structured using date-based features.

4. **Data Accuracy**

Temperature and wind speed values fall within realistic environmental ranges.

The dataset contains both rainy and clear days, indicating natural variability in weather conditions.

Overall, the dataset is well-structured, clean, and suitable for statistical analysis, visualization, and machine learning applications.

I’ve been working on a **Weather Data Analysis project using Python and Pandas**, and it has been a great learning experience so far.

The dataset contains **428 daily weather records**, including variables such as temperature, precipitation, wind speed, and seasonal information. Through this project, I explored how raw data can be transformed into meaningful insights.

Some of the key work done so far includes:
• Data cleaning and preprocessing
• Descriptive statistical analysis
• Identifying hot days, windy days, and extreme weather conditions
• Seasonal pattern exploration
• Dataset quality evaluation and feature creation

This project is still **work in progress**, and I plan to keep improving it by adding:
• More advanced data visualizations
• Deeper exploratory data analysis (EDA)
• Trend analysis and pattern detection
• Possibly building a simple predictive model in the next phase

It’s exciting to see how structured analysis can reveal patterns hidden inside everyday data like weather records.

Always open to feedback and suggestions for improvement as I continue refining this project.

#DataAnalytics #Python #Pandas #DataScience #EDA #WeatherData #LearningInPublic #AnalyticsJourney
